File: go-backend/internal/dto/blog.go
package dto

import (
	"time"

	"github.com/KanoCifer/kuroome-blog/internal/mongo/document"
)

// PostResponse 单篇博客输出 —— 与 Python _serialize_post 形状一致。
type PostResponse struct {
	ID        string   `json:"_id"`
	Title     string   `json:"title"`
	Body      string   `json:"body"`
	Summary   *string  `json:"summary"`
	Cover     *string  `json:"cover"`
	Tags      []string `json:"tags"`
	IsPinned  int      `json:"is_pinned"`
	Views     int      `json:"views"`
	Likes     int      `json:"likes"`
	CreatedAt string   `json:"created_at"`
	UpdatedAt string   `json:"updated_at"`
}

// LikeResponse 点赞响应 —— 返回递增后的最新喜欢数。
type LikeResponse struct {
	Likes int `json:"likes"`
}

// TagResponse 标签聚合项 —— 与 Python aggregate_tag_counts 形状一致。
type TagResponse struct {
	Name  string `json:"name"`
	Count int    `json:"count"`
}

// BlogListResponse 博客列表响应 —— 与 Python get_blogs 返回形状一致。
type BlogListResponse struct {
	Posts      []PostResponse `json:"posts"`
	Tags       []TagResponse  `json:"tags"`
	Pagination Pagination     `json:"pagination"`
}

// PostsByTagResponse 标签筛选响应 —— 与 Python get_posts_by_tag 返回形状一致。
type PostsByTagResponse struct {
	Posts []PostResponse `json:"posts"`
	Tag   string         `json:"tag"`
	Pagination Pagination     `json:"pagination"`
}

// ToPostResponse document → 输出 DTO —— 与 Python _serialize_post 对齐。
// mongo-driver 将 _id(ObjectID) 解码为 ID 字段的十六进制字符串。
func ToPostResponse(p document.Post) PostResponse {
	return PostResponse{
		ID:        p.ID,
		Title:     p.Title,
		Body:      p.Body,
		Summary:   p.Summary,
		Cover:     p.Cover,
		Tags:      p.Tags,
		IsPinned:  p.IsPinned,
		Views:     p.Views,
		Likes:     p.Likes,
		CreatedAt: formatTime(p.CreatedAt),
		UpdatedAt: formatTime(p.UpdatedAt),
	}
}

// ToPostList 批量转换，使用 ToResponseSlice 统一实现。
func ToPostList(items []document.Post) []PostResponse {
	return ToResponseSlice(items, ToPostResponse)
}

// formatTime 将 time.Time 格式化为 RFC3339 字符串；零值返回 ""。
func formatTime(tm time.Time) string {
	if tm.IsZero() {
		return ""
	}
	return tm.UTC().Format(time.RFC3339)
}

File: packages/types/src/blog.ts
// ── 博客文章类型 ────────────────────────────────────────────────────────────
// Post = PostgreSQL 存储（id 主键），BlogPost = MongoDB 存储（_id 主键）

export interface Post {
  id?: number;
  _id?: string;
  title: string;
  content?: string;
  body: string;
  summary?: string;
  cover?: string;
  author_id?: number;
  author_name?: string;
  author?: string;
  views?: number;
  likes?: number;
  created_at: string;
  updated_at?: string;
  tags?: string[];
  is_pinned?: boolean;
}

export interface BlogPost {
  _id: string;
  title: string;
  body: string;
  summary?: string | null;
  cover?: string | null;
  views?: number;
  likes?: number;
  created_at: string;
  updated_at: string;
  tags: string[];
  is_pinned?: boolean;
}

/** 标签聚合项（带文章计数） */
export interface TagItem {
  name: string;
  count: number;
}

/** 标签筛选结果 */
export interface PostsByTagResponse {
  posts: BlogPost[];
  tag: string;
  total: number;
}

export interface BlogPagination {
  page: number;
  per_page: number;
  total: number;
  pages: number;
  has_prev: boolean;
  has_next: boolean;
  prev_num?: number | null;
  next_num?: number | null;
}

/** 博客列表查询参数 */
export interface BlogQuery {
  page?: number;
  search?: number;
}

/** 博客列表响应（v1/v2 旧格式，含 tags 聚合） */
export interface BlogsResponse {
  status: string;
  message: string;
  data: {
    posts: Post[];
    tags: TagItem[];
    pagination: BlogPagination;
  };
}

/** 博客列表响应（v3 新格式） */
export interface BlogListResponse {
  posts: BlogPost[];
  tags: TagItem[];
  pagination: BlogPagination;
}

/** 单篇文章响应（v3 格式，含 is_pinned / views / likes） */
export interface BlogPostResponse {
  _id: string;
  title: string;
  body: string;
  summary?: string | null;
  cover?: string | null;
  tags: string[];
  is_pinned: boolean;
  views?: number;
  likes?: number;
  created_at: string;
  updated_at: string;
}

export interface PostResponse {
  status: string;
  message: string;
  data: Post;
}
